### ISSUE TITLE
Closure in combination with decorators incorrectly assumes functions are not used

### ISSUE THREAD

**Issue by remcovaes** [microsoft/pyright#6197]
**Describe the bug**
When using Flask and passing the app into functions to register endpoints, you will run into this. The code sample provides a minimal reproduction. When using decorators in a closure, it is incorrectly assumed that the decorated function in the closure is not used.

This happens in strict mode.

**Code**
```python
from typing import Any


class Foo:
    def some_decorator(self, fn: Any):
        fn()


def fn(foo: Foo):
    @foo.some_decorator
    def bar():
        print("bar")


fn(Foo())
```
**Expected**
No error as `bar` is used

**Actual**
Error that bar is not used

**VS Code extension or command-line**
Confirmed bug on both

**Comment by erictraut** [comment]
The behavior is correct. Your code is equivalent to:
```python
def fn(foo: Foo):
    def _bar():
        print("bar")

    bar = foo.some_decorator(_bar)
```

The symbol `bar` is not being used. If you enable `reportUnusedVariable` (which is on by default in strict mode), this will generate an error. If you don't want to see errors like this, you can switch to "basic" type checking mode and enable only a subset of diagnostic checks that are included in strict mode. Alternatively, you can use a `# pyright: ignore[reportUnusedVariable]` comment at the end of the line or use a `# pyright: reportUnusedVariable=false` at the top of the file.

**Comment by remcovaes** [comment]
Running the code results in bar being printed, which means that the function is used. The underscore is a good workaround indeed.

**Comment by remcovaes** [comment]
The code example is a simplified version of what flask would do with endpoints that are decorated. I agree that the decorator is a bit weird. I will use the underscore for my flask endpoints. 

Thanks for the explanation!
